Have anyhow::Error implement std::error::Error?

`anyhow::Error` does not implement `std::error::Error`, which leads to confusion when using anyhow with code that expects a `Result<_, std::error::Error>`.

I noticed that the same applies to `failure::Error` in the failure crate.

What is the reason for that? What prevents `anyhow::Error` from implementing `std::error::Error`?

... and thanks for all your awesome Rust crates!

That impl would be incoherent with the existing pair of impls `impl From<T> for T` (libcore) and `impl<E> From<E> for anyhow::Error where E: std::error::Error + Send + Sync + 'static` (`?` support).

```console
error[E0119]: conflicting implementations of trait `std::convert::From<anyhow::Error>` for type `anyhow::Error`:
   --> src/error.rs:313:1
    |
313 | / impl<E> From<E> for Error
314 | | where
315 | |     E: std::error::Error + Send + Sync + 'static,
316 | | {
...   |
320 | |     }
321 | | }
    | |_^
    |
    = note: conflicting implementation in crate `core`:
            - impl<T> std::convert::From<T> for T;
```

Can you give a minimal example of code that would want this impl to exist?

--- a/src/error.rs
+++ b/src/error.rs
@@ -572,6 +572,12 @@ impl From<Error> for Box<dyn StdError + Send + Sync + 'static> {
     }
 }
 
+impl From<Error> for Box<dyn StdError + 'static> {
+    fn from(error: Error) -> Self {
+        Box::<dyn StdError + Send + Sync>::from(error)
+    }
+}
+
 /// Iterator of a chain of source errors.
 ///
 /// This type is the iterator returned by [`Error::chain`].

--- a/tests/test_convert.rs
+++ b/tests/test_convert.rs
@@ -1,7 +1,7 @@
 mod drop;
 
 use self::drop::DetectDrop;
-use anyhow::Error;
+use anyhow::{Error, Result};
 use std::error::Error as StdError;
 use std::sync::atomic::AtomicBool;
 use std::sync::atomic::Ordering::SeqCst;
@@ -15,4 +15,13 @@ fn test_convert() {
     assert_eq!("oh no!", box_dyn.to_string());
     drop(box_dyn);
     assert!(has_dropped.load(SeqCst));
+}
+
+#[test]
+fn test_question_mark() -> Result<(), Box<dyn StdError>> {
+    fn f() -> Result<()> {
+        Ok(())
+    }
+    f()?;
+    Ok(())
 }

